src/encoder.js

```diff
@@ -270,41 +271,100 @@ export default class Encoder extends Extendable {
+                extTarget.push(tags.EXTENSION_);
+                encodeUInt(extensionPoint, extTarget);
+              }
+              let furtherEnabledExtensions = applyRecursively
               ? [...enabledExtensions, extensionPoint]
               : enabledExtensions;
-            this.encodeValue(extension.serialise(placeholder.value), extTarget, furtherEnabledExtensions);
-          } else {
-            this.encodeValue(placeholder.value, extTarget, enabledExtensions);
+              this.encodeValue(extension.serialise(placeholder.value), extTarget, furtherEnabledExtensions);
+            } else {
+              this.encodeValue(placeholder.value, extTarget, enabledExtensions);
+            }
+            replacements.push({ index: placeholder.location, data: extTarget, placeholders: this.placeholderMap });
           }
-          target.splice(target.indexOf(placeholder), 1, ...extTarget);
         }
       }
 
-      if (typeof extension.memo === 'function') {
-        target.unshift(...this.encodeValue(extension.memo(), [], enabledExtensions));
+      this.placeholderMap = {};
+
+      let returnData = [];
+      let lastIndex = 0;
+
+      // restructure data into new array
+      for (let replacement of replacements) {
+        if (replacement.index < lastIndex) {
+          throw new Error('Corrupt non-linear state');
+        } else if (replacement.index > lastIndex) {
+          returnData.push(...target.slice(lastIndex, replacement.index)); // skips placeholder byte
+        }
+        for (let extensionPoint in replacement.placeholders) {
+          if (!(extensionPoint in this.placeholderMap)) {
+            this.placeholderMap[extensionPoint] = [];
+          }
+          let placeholders = replacement.placeholders[extensionPoint];
+          for (let placeholder of placeholders) {
+            this.placeholderMap[extensionPoint].push({ value: placeholder.value, extensionPoint: extensionPoint, location: placeholder.location + returnData.length });
+          }
+        }
+
+        // fill in replacement data
+        returnData.push(...replacement.data);
+        lastIndex = replacement.index + 1;
       }
+      
+      returnData.push(...target.slice(lastIndex));
+      target = returnData;
     }
 
+    if (memo) { // build memos in series
+      let enabledExtensions = extensions.slice(0);
+      let prepend = [];
+      
+      while (enabledExtensions.length > 0) {
+        let extensionPoint = enabledExtensions.pop();
+        let extension = this.extensions[extensionPoint];
+        if (typeof extension.memo === 'function') {
+          this.placeholderMap = {};
+          let extTarget = [];
+          let predata = this.encodeValue(extension.memo(), extTarget, enabledExtensions);
+          let memoData = this.replaceExtensionPlaceholders(predata, enabledExtensions, false);
+          prepend.push(memoData);
+        }
+      }
+      // combine them all together
+      let finalData = [];
```

Comment:
I think we can replace this whole section with a multiple-parameter `[].concat` call.

Reply:
I don't think we can reverse the contents of the `prepend` array with a concat call, unless we also call `reverse` on `prepend`. I don't see a problem with that myself, alternatively we can leave it as is.

Reply:
I guess we could also just `target.unshift(...[].concat.call([], prepend))`?

Reply:
We could, but unshift is not ideal for performance sensitive code. Should I do it anyways?